### PR TITLE
Issue/v4/984 properties def

### PR DESCRIPTION
**Pull request by stefanorosanelli**
This PR solves #984 
`/properties` endpoint is introduced and relationships with `object_types` is handled

How to use `properties` will be addressed in a future issue
Also what initial DB data will be needed

### CODE REVIEW

File: plugins/BEdita/API/src/Utility/JsonApi.php [hunk 1]
```diff
@@ -130,8 +132,9 @@ function (BelongsToMany $val) {
 
         foreach ($associations as $association) {
             list(, $type) = namespaceSplit(get_class($association));
-            $name = $association->property();
+            $name = Inflector::underscore($association->name());
```
Comment (fquffio) [inline plugins/BEdita/API/src/Utility/JsonApi.php]: Uhm… Why did you change this? Association name might not be the same as the entity's populated property. E.g.:

```
// In a Table's `initialize()` method:
$this->belongsTo('Foos', [
    'propertyName' => 'my_foo_property_name',
]);
```

Also, with "hasOne" associations this will probably fail even with a non-custom configuration, as the property name is in the singular form.

File: plugins/BEdita/API/src/Controller/ResourcesController.php [hunk 2]
```diff
@@ -96,6 +99,9 @@ public function relationships()
         $id = $this->request->param('id');
         $relationship = $this->request->param('relationship');
 
+        $allowAssoc = Inflector::underscore($this->name);
+        $this->set(['_allowedAssociations' => [ $allowAssoc => [$allowAssoc]]]);
```
Comment (fquffio) [inline plugins/BEdita/API/src/Controller/ResourcesController.php]: Honestly I don't really like this so much. 😜 It works, but controller name could differ from reverse association's name, especially in objects controller, where we'll have to deal with several models.

I think that this "allowed associations" property should be refactored and exposed at the Entity or Table level. But this could be done in the future. For now, this could probably be changed to something like:

```
// Put these lines just below "$Association = $this->findAssociation($relationship);"

// Try to guess reverse association and implicitly add it to displayed associations.
$reverseAssociation = $Association->target()->association($this->modelClass);
if ($reverseAssociation !== null) {
    $allowAssoc = $reverseAssociation->property();
    $this->set(['_allowedAssociations' => [$allowAssoc => [$allowAssoc]]]);
}
```

**Review by fquffio — COMMENTED** [review]
I added a couple of notes regarding `allowedAssociations` handling: in both cases things do work fine, so it's more like a personal taste, nothing critical at all.

Also, I'd proceed with a merge of `4-corylus` in this branch before merging this PR, as I fear that a schema migration should be baked to port schema changes applied in this branch.

File: plugins/BEdita/API/src/Utility/JsonApi.php (hunk 1 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/src/Utility/JsonApi.php]: Ok, committed

File: plugins/BEdita/API/src/Controller/ResourcesController.php (hunk 2 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/src/Controller/ResourcesController.php]: Sometimes we would probably want to hide some entity/table associations from controller response, we'll think about that later.
Reverse model associations committed.

**Comment by stefanorosanelli** [comment]
Branch `4-develop` merged in issue branch and migrations updated.
